File: lib/dns-resolve.c
/*
 * Copyright (c) 2017, 2018 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>
#include "dns-resolve.h"
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <arpa/nameser.h>
#include <errno.h>
#include <string.h>
#include <sys/stat.h>
#include <unbound.h>
#include "hash.h"
#include "openvswitch/hmap.h"
#include "openvswitch/vlog.h"
#include "timeval.h"

VLOG_DEFINE_THIS_MODULE(dns_resolve);

/* Guard all_reqs__ and resolve_state of each request. */
static struct ovs_mutex dns_mutex__ = OVS_MUTEX_INITIALIZER;
static struct hmap all_reqs__;
static struct ub_ctx *ub_ctx__;

static bool thread_is_daemon;

static struct vlog_rate_limit rl = VLOG_RATE_LIMIT_INIT(1, 1);

enum resolve_state {
    RESOLVE_INVALID,
    RESOLVE_PENDING,
    RESOLVE_GOOD,
    RESOLVE_ERROR
};

struct resolve_request {
    struct hmap_node hmap_node;     /* node for all_reqs__ */
    char *name;                     /* the domain name to be resolved */
    char *addr;                     /* the resolved ip address */
    enum resolve_state state;       /* state of this request */
    time_t time;                    /* resolving time */
    struct ub_result *ub_result;    /* the stored unbound result */
};

static struct resolve_request *resolve_find_or_new__(const char *name)
    OVS_REQUIRES(dns_mutex__);
static bool resolve_check_expire__(struct resolve_request *req)
    OVS_REQUIRES(dns_mutex__);
static bool resolve_check_valid__(struct resolve_request *req)
    OVS_REQUIRES(dns_mutex__);
static bool resolve_async__(struct resolve_request *req, int qtype)
    OVS_REQUIRES(dns_mutex__);
static void resolve_callback__(void *req, int err, struct ub_result *)
    OVS_REQUIRES(dns_mutex__);
static bool resolve_result_to_addr__(struct ub_result *result, char **addr);
static bool dns_resolve_sync__(const char *name, char **addr);

/* Pass a true 'is_daemon' if you don't want the DNS-resolving to block the
 * running thread.
 */
void
dns_resolve_init(bool is_daemon)
{
    ub_ctx__ = ub_ctx_create();
    if (ub_ctx__ == NULL) {
        VLOG_ERR_RL(&rl, "Failed to create libunbound context, "
        "so asynchronous DNS resolving is disabled.");
        return;
    }

    const char *ub_conf_filename = getenv("OVS_UNBOUND_CONF");
    if (ub_conf_filename != NULL) {
        int retval = ub_ctx_config(ub_ctx__, ub_conf_filename);
        if (retval != 0) {
            VLOG_WARN_RL(&rl, "Failed to set libunbound context config: %s",
                         ub_strerror(retval));
            ub_ctx_delete(ub_ctx__);
            ub_ctx__ = NULL;
            return;
        }
    }

    const char *filename = getenv("OVS_RESOLV_CONF");
    if (!filename) {
#ifdef _WIN32
        /* On Windows, NULL means to use the system default nameserver. */
#else
        filename = "/etc/resolv.conf";
#endif
    }
    struct stat s;
    if (!filename || !stat(filename, &s) || errno != ENOENT) {
        int retval = ub_ctx_resolvconf(ub_ctx__, filename);
        if (retval != 0) {
            VLOG_WARN_RL(&rl, "Failed to read %s: %s",
                         filename ? filename : "system default nameserver",
                         ub_strerror(retval));
            ub_ctx_delete(ub_ctx__);
            ub_ctx__ = NULL;
            return;
        }
    } else {
        VLOG_WARN_RL(&rl, "Failed to read %s: %s",
                     filename, ovs_strerror(errno));
        ub_ctx_delete(ub_ctx__);
        ub_ctx__ = NULL;
        return;
    }

    /* Handles '/etc/hosts' on Linux and 'WINDIR/etc/hosts' on Windows. */
    int retval = ub_ctx_hosts(ub_ctx__, NULL);
    if (retval != 0) {
        VLOG_WARN_RL(&rl, "Failed to read etc/hosts: %s",
                     ub_strerror(retval));
    }

    ub_ctx_async(ub_ctx__, true);
    hmap_init(&all_reqs__);
    thread_is_daemon = is_daemon;
}

/* Returns true on success. Otherwise, returns false and the error information
 * can be found in logs. If there is no error information, then the resolving
 * is in process and the caller should call again later. The value of '*addr'
 * is always nullified if false is returned. If this function is called under
 * daemon-context, the resolving will undergo asynchronously. Otherwise, a
 * synchronouse resolving will take place.
 *
 * This function is thread-safe.
 *
 * The caller is responsible for freeing the returned '*addr'.
 */
bool
dns_resolve(const char *name, char **addr)
    OVS_EXCLUDED(dns_mutex__)
{
    bool success = false;

    if (!thread_is_daemon) {
        return dns_resolve_sync__(name, addr);
    }

    *addr = NULL;
    ovs_mutex_lock(&dns_mutex__);

    if (ub_ctx__ == NULL) {
        goto unlock;
    }

    /* ub_process is inside lock as it invokes resolve_callback__. */
    int retval = ub_process(ub_ctx__);
    if (retval != 0) {
        VLOG_ERR_RL(&rl, "dns-resolve error: %s", ub_strerror(retval));
        goto unlock;
    }

    struct resolve_request *req;
    req = resolve_find_or_new__(name);
    if (resolve_check_valid__(req)) {
        *addr = xstrdup(req->addr);
        success = true;
    } else if (req->state != RESOLVE_PENDING) {
        success = resolve_async__(req, ns_t_a);
    }
unlock:
    ovs_mutex_unlock(&dns_mutex__);
    return success;
}

void
dns_resolve_destroy(void)
{
    if (ub_ctx__ != NULL) {
        /* Outstanding requests will be killed. */
        ub_ctx_delete(ub_ctx__);
        ub_ctx__ = NULL;

        struct resolve_request *req, *next;
        HMAP_FOR_EACH_SAFE (req, next, hmap_node, &all_reqs__) {
            ub_resolve_free(req->ub_result);
            free(req->addr);
            free(req->name);
            free(req);
        }
        hmap_destroy(&all_reqs__);
    }
}

static struct resolve_request *
resolve_find_or_new__(const char *name)
    OVS_REQUIRES(dns_mutex__)
{
    struct resolve_request *req;

    HMAP_FOR_EACH_IN_BUCKET(req, hmap_node, hash_string(name, 0),
                            &all_reqs__) {
        if (!strcmp(name, req->name)) {
            return req;
        }
    }

    req = xzalloc(sizeof *req);
    req->name = xstrdup(name);
    req->state = RESOLVE_INVALID;
    hmap_insert(&all_reqs__, &req->hmap_node, hash_string(req->name, 0));
    return req;
}

static bool
resolve_check_expire__(struct resolve_request *req)
    OVS_REQUIRES(dns_mutex__)
{
    return time_now() > req->time + req->ub_result->ttl;
}

static bool
resolve_check_valid__(struct resolve_request *req)
    OVS_REQUIRES(dns_mutex__)
{
    return (req != NULL
        && req->state == RESOLVE_GOOD
        && !resolve_check_expire__(req));
}

static bool
resolve_async__(struct resolve_request *req, int qtype)
    OVS_REQUIRES(dns_mutex__)
{
    if (qtype == ns_t_a || qtype == ns_t_aaaa) {
        int retval;
        retval = ub_resolve_async(ub_ctx__, req->name,
                                  qtype, ns_c_in, req,
                                  resolve_callback__, NULL);
        if (retval != 0) {
            req->state = RESOLVE_ERROR;
            return false;
        } else {
            req->state = RESOLVE_PENDING;
            return true;
        }
    }
    return false;
}

static void
resolve_callback__(void *req_, int err, struct ub_result *result)
    OVS_REQUIRES(dns_mutex__)
{
    struct resolve_request *req = req_;

    if (err != 0 || (result->qtype == ns_t_aaaa && !result->havedata)) {
        ub_resolve_free(result);
        req->state = RESOLVE_ERROR;
        VLOG_ERR_RL(&rl, "%s: failed to resolve", req->name);
        return;
    }

    /* IPv4 address is empty, try IPv6. */
    if (result->qtype == ns_t_a && !result->havedata) {
        ub_resolve_free(result);
        resolve_async__(req, ns_t_aaaa);
        return;
    }

    char *addr;
    if (!resolve_result_to_addr__(result, &addr)) {
        ub_resolve_free(result);
        req->state = RESOLVE_ERROR;
        VLOG_ERR_RL(&rl, "%s: failed to resolve", req->name);
        return;
    }

    ub_resolve_free(req->ub_result);
    free(req->addr);

    req->ub_result = result;
    req->addr = addr;
    req->state = RESOLVE_GOOD;
    req->time = time_now();
}

static bool
resolve_result_to_addr__(struct ub_result *result, char **addr)
{
    int af = result->qtype == ns_t_a ? AF_INET : AF_INET6;
    char buffer[INET6_ADDRSTRLEN];

    /* XXX: only the first returned IP is used. */
    if (inet_ntop(af, result->data[0], buffer, sizeof buffer)) {
        *addr = xstrdup(buffer);
    } else {
        *addr = NULL;
    }

    return (*addr != NULL);
}

static bool
dns_resolve_sync__(const char *name, char **addr)
{
    *addr = NULL;

    if (ub_ctx__ == NULL) {
        dns_resolve_init(false);
        if (ub_ctx__ == NULL) {
            return false;
        }
    }

    struct ub_result *result;
    int retval = ub_resolve(ub_ctx__, name, ns_t_a, ns_c_in, &result);
    if (retval != 0) {
        return false;
    } else if (!result->havedata) {
        ub_resolve_free(result);

        retval = ub_resolve(ub_ctx__, name, ns_t_aaaa, ns_c_in, &result);
        if (retval != 0) {
            return false;
        } else if (!result->havedata) {
            ub_resolve_free(result);
            return false;
        }
    }

    bool success = resolve_result_to_addr__(result, addr);
    ub_resolve_free(result);
    return success;
}

File: python/ovs/dns_resolve.py
# Copyright (c) 2023 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import collections
import enum
import functools
import ipaddress
import os
import time
import typing

try:
    import unbound  # type: ignore
except ImportError:
    pass

import ovs.vlog

vlog = ovs.vlog.Vlog("dns_resolve")

RESOLVE_UNUSED_TIMEOUT = 300  # Seconds.


class ReqState(enum.Enum):
    INVALID = 0
    PENDING = 1
    GOOD_UNUSED = 2
    GOOD = 3
    ERROR = 4


class DNSRequest:
    def __init__(self, name: str):
        self.name: str = name
        self.state: ReqState = ReqState.INVALID
        self.time: typing.Optional[float] = None
        # set by DNSResolver._callback
        self.result: typing.Optional[str] = None
        self.ttl: typing.Optional[float] = None

    @property
    def expired(self):
        ttl = self.ttl
        # When we just sent a request, but didn't look at the response yet,
        # it's not caching, but a "transaction in progress" situation, so we
        # can use the response even with TTL of 0 and more than 1 second
        # passed.  Allow such values to be accessed for at least
        # RESOLVE_UNUSED_TIMEOUT seconds without considering them stale.
        # This is necessary in case of large backoff intervals on connections
        # or if the process is doing some other work not looking at the
        # response for longer than TTL.
        if self.state == ReqState.GOOD_UNUSED:
            ttl = max(ttl, RESOLVE_UNUSED_TIMEOUT)
            # Not a "transaction in progress" anymore, normal caching rules
            # should apply from this point forward.
            self.state = ReqState.GOOD

        return time.time() > self.time + ttl

    @property
    def is_valid(self):
        return (self.state in [ReqState.GOOD_UNUSED, ReqState.GOOD]
                and not self.expired)

    def __str__(self):
        return (f"DNSRequest(name={self.name}, state={self.state}, "
                f"time={self.time}, result={self.result})")


class DefaultReqDict(collections.defaultdict):
    def __init__(self):
        super().__init__(DNSRequest)

    def __missing__(self, key):
        ret = self.default_factory(key)
        self[key] = ret
        return ret


class UnboundException(Exception):
    def __init__(self, message, errno):
        try:
            msg = f"{message}: {unbound.ub_strerror(errno)}"
        except NameError:
            msg = message
        super().__init__(msg)


def dns_enabled(func):
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        if self.dns_enabled:
            return func(self, *args, **kwargs)
        vlog.err("DNS support requires the python unbound library")
    return wrapper


class DNSResolver:
    def __init__(self, is_daemon: bool = False):
        """Create a resolver instance

        If is_daemon is true, set the resolver to handle requests
        asynchronously. The following environment variables are processed:

        OVS_UNBOUND_CONF: The filename for an unbound.conf file
        OVS_RESOLV_CONF: A filename to override the system default resolv.conf
        OVS_HOSTS_FILE: A filename to override the system default hosts file

        In the event that the unbound library is missing or fails to initialize
        DNS lookup support will be disabled and the resolve() method will
        return None.
        """
        self._is_daemon = is_daemon
        try:
            self._ctx = unbound.ub_ctx()
            self.dns_enabled = True
        except Exception:
            # The unbound docs mention that this could thrown an exception
            # but do not specify what exception that is. This can also
            # happen with a missing unbound library.
            self.dns_enabled = False
            vlog.err("Failed to initialize the unbound library")
            return

        # NOTE(twilson) This cache, like the C version, can grow without bound
        # and has no cleanup or aging mechanism. Given our usage patterns, this
        # should not be a problem. But this should not be used to resolve an
        # unbounded list of addresses in a long-running daemon.
        self._requests = DefaultReqDict()

        self._ub_call(self._set_unbound_conf)

        # NOTE(twilson) The C version disables DNS in this case. I didn't do
        # that here since it could still be useful to resolve addresses from
        # /etc/hosts even w/o resolv.conf
        self._ub_call(self._set_resolv_conf)
        self._ub_call(self._set_hosts_file)

        self._ctx.set_async(True)  # Sets threaded behavior for resolve_async()

    def _ub_call(self, fn, *args, **kwargs):
        """Convert UnboundExceptions into vlog warnings"""
        try:
            return fn(*args, **kwargs)
        except UnboundException as e:
            vlog.warn(e)

    @dns_enabled
    def _set_unbound_conf(self):
        ub_cfg = os.getenv("OVS_UNBOUND_CONF")
        if ub_cfg:
            retval = self._ctx.config(ub_cfg)
            if retval != 0:
                raise UnboundException(
                    "Failed to set libunbound context config", retval)

    @dns_enabled
    def _set_resolv_conf(self):
        filename = os.getenv("OVS_RESOLV_CONF")
        # The C lib checks that the file exists and also sets filename to
        # /etc/resolv.conf on non-Windows, but resolvconf already does this.
        retval = self._ctx.resolvconf(filename)
        if retval != 0:
            location = filename or "system default nameserver"
            raise UnboundException(location, retval)

    @dns_enabled
    def _set_hosts_file(self):
        # The C lib doesn't have the ability to set a hosts file, but it is
        # useful to have, especially for writing tests that don't rely on
        # network connectivity. hosts(None) uses /etc/hosts.
        filename = os.getenv("OVS_HOSTS_FILE")
        retval = self._ctx.hosts(filename)
        if retval != 0:
            location = filename or "system default hosts file"
            raise UnboundException(location, retval)

    @dns_enabled
    def _callback(self, req: DNSRequest, err: int, result):
        if err != 0 or (result.qtype == unbound.RR_TYPE_AAAA
                        and not result.havedata):
            req.state = ReqState.ERROR
            vlog.warn(f"{req.name}: failed to resolve")
            return
        if result.qtype == unbound.RR_TYPE_A and not result.havedata:
            self._resolve_async(req, unbound.RR_TYPE_AAAA)
            return
        try:
            ip_str = next(iter(result.data.as_raw_data()))
            ip = ipaddress.ip_address(ip_str)  # test if IP is valid
            # NOTE (twilson) For some reason, accessing result data outside of
            # _callback causes a segfault. So just grab and store what we need.
            req.result = str(ip)
            req.ttl = result.ttl
            req.state = ReqState.GOOD_UNUSED
            req.time = time.time()
        except (ValueError, StopIteration):
            req.state = ReqState.ERROR
            vlog.err(f"{req.name}: failed to resolve")

    @dns_enabled
    def _resolve_sync(self, name: str) -> typing.Optional[str]:
        for qtype in (unbound.RR_TYPE_A, unbound.RR_TYPE_AAAA):
            err, result = self._ctx.resolve(name, qtype)
            if err != 0:
                return None
            if not result.havedata:
                continue
            try:
                ip = ipaddress.ip_address(
                    next(iter(result.data.as_raw_data())))
            except (ValueError, StopIteration):
                return None
            return str(ip)

        return None

    @dns_enabled
    def _resolve_async(self, req: DNSRequest, qtype) -> None:
        err, _ = self._ctx.resolve_async(req.name, req, self._callback,
                                         qtype)
        if err != 0:
            req.state = ReqState.ERROR
            return None

        req.state = ReqState.PENDING
        return None

    @dns_enabled
    def resolve(self, name: str) -> typing.Optional[str]:
        """Resolve a host name to an IP address

        If the resolver is set to handle requests asynchronously, resolve()
        should be recalled until it returns a non-None result. Errors will be
        logged.

        :param name: The host name to resolve
        :returns: The IP address or None on error or not (yet) found
        """
        if not self._is_daemon:
            return self._resolve_sync(name)
        retval = self._ctx.process()
        if retval != 0:
            vlog.err(f"dns-resolve error: {unbound.ub_strerror(retval)}")
            return None
        req = self._requests[name]  # Creates a DNSRequest if not found
        if req.is_valid:
            return req.result
        elif req.state != ReqState.PENDING:
            self._resolve_async(req, unbound.RR_TYPE_A)
        return None


_global_resolver: typing.Optional[DNSResolver] = None


def init(is_daemon: bool = False) -> DNSResolver:
    """Initialize a global DNSResolver

    See DNSResolver.__init__ for more details
    """
    global _global_resolver
    _global_resolver = DNSResolver(is_daemon)
    return _global_resolver


def resolve(name: str) -> typing.Optional[str]:
    """Resolve a host name to an IP address

    If a DNSResolver instance has not been instantiated, or if it has been
    created with is_daemon=False, resolve() will synchronously resolve the
    hostname. If DNSResolver has been initialized with is_daemon=True, it
    will instead resolve asynchornously and resolve() will return None until
    the hostname has been resolved.

    :param name: The host name to resolve
    :returns: The IP address or None on error or not (yet) found
    """
    if _global_resolver is None:
        init()

    # mypy doesn't understand that init() sets _global_resolver, so ignore type
    return _global_resolver.resolve(name)  # type: ignore


def destroy():
    """Destroy the global DNSResolver

    This destroys the global DNSResolver instance and any outstanding
    asynchronouse requests.
    """
    global _global_resolver
    del _global_resolver
    _global_resolver = None  # noqa: F841
